x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -983,8 +984,10 @@ public String toString() {
             // conditions on the `principal` and `realm` fields
             // 2. reusability of the AuditEventMetaInfo instance: in this case Streams have
             // to be regenerated as they cannot be operated upon twice
-            this.roles = () -> roles.filter(r -> r.length != 0).map(Arrays::stream).orElse(Stream.of(""));
-            this.indices = () -> indices.filter(i -> i.length != 0).map(Arrays::stream).orElse(Stream.of(""));
+            this.roles = () -> roles.filter(r -> r.length > 0).filter(r -> r.length > 1 || r[0] != null).map(Arrays::stream)
+                    .orElse(Stream.of(""));
+            this.indices = () -> indices.filter(i -> i.length > 0).filter(i -> i.length > 1 || i[0] != null).map(Arrays::stream)
+                    .orElse(Stream.of(""));
```

Comment:
This is fine, but I think I would be confused by it if I hadn't already read the PR description.
I think something like this is more readable?
```
   opt.filter(a -> Arrays.stream(a).anyMatch(Objects::nonNull)).map(Arrays::stream).orElse(Stream.of(""));
```
It's slightly different, in that it will treat `{ null, null, null }` (etc) as "empty", but I think that's OK.

Reply:
preciate Tim, I went on with your suggestion!